### ISSUE TITLE
Need some advice for reproducing the results & opinion about RRS sampling

### ISSUE THREAD

**Issue by HJoonKwon** [AsuradaYuci/TF-CLIP#13]
Thanks for sharing the great work and the entire code base!

I'm trying to reproduce the TF-CLIP's performance on MARS by training it in my training pipeline. 

On MARS, the mAP is somewhat similar (~0.885), but the rank-1 (~0.913) is much worse than the reported number. 

<img width="687" height="292" alt="Image" src="https://github.com/user-attachments/assets/2b60c610-1739-4819-bf0d-90778ea22d3b" />

What was the most sensitive factor while training the model? 

Things that can differ:

1) attention computations (I'm using FlashAttention implemented in pytorch)
2) randomness in the sampling process (both in sampling pids and frames out of tracklets)
3) loss computations (I'm using [pytorch-metric-learning](https://github.com/KevinMusgrave/pytorch-metric-learning) for triplet loss)
4) I'm not using SIE

Regarding the RRS sampling, 

Do you think the model can capture the "temporal" information based on this sampling method? 
My intuition is that this method does not sample consecutive(temporal) frames, but more like 'diverse' frames of each tracklet. If a tracklet is long enough, I guess that the temporal correlation among sampled frames would be very low. Also, the model is using full-attention across the temporal dimension, which means the model would not care even if we just randomly shuffle the sampled frames. 

Sorry for a lot of questions on one issue. 

Thank you!

**Comment by AsuradaYuci** [comment]
@HJoonKwon （1）I haven't tried training with FlashAttention myself, so I'm not sure about its specific impact.
Regarding the other factors:
         SIE: Adding SIE usually brings an improvement of about 0.3%.
         Sampling: Compared to random sampling, the RRS (Restricted Random Sampling) method typically yields a gain of around 0.5% on ViT/ResNet backbones. However, I haven't conducted relevant experiments on CLIP-based models.
         Loss: The influence of the loss function implementation shouldn't be very significant.
（2）**Regarding the real purpose of RRS:**
The original design intention of RRS is not to capture temporal motion information (i.e., "the action of the person walking").
Instead, its purpose is to capture appearance diversity. A person in a video might be occluded or might turn around. RRS attempts to select frames with greater differences within a long tracklet so that the model can piece together the person's complete appearance features (e.g., front, side, and back views).

**Comment by HJoonKwon** [comment]
@AsuradaYuci 

Thank you for the detailed answer! I think FlashAttention should not have any visible impact.

Have a following question: 
1) For SIE, 0.3% improvement means the absolute value or relative value? (like 0.90 + 0.03= 0.903 or 0.90*1.03?=0.927)
2) I was exploring some potential applications of this approach in robotics. But, it seems like the purpose of RRS sampling is more on the offline people searching rather than real-time robotics applications? Because we need a complete tracklet (ideally long enough) and run RRS sampling to get a 'diverse' clip for that specific person.

**Comment by AsuradaYuci** [comment]
@HJoonKwon 
1. For SIE, 0.3% improvement means the absolute value (0.90 + 0.03= 0.903).
2.  If you want to utilize all available frames, there is actually a 'dense' version of RRS. Alternatively, during testing, you can change `sampler_method = ['rrs_test'`](https://github.com/AsuradaYuci/TF-CLIP/blob/497a4f24eee1385fc3b20eaaceb0aadd75ce851f/datasets/make_dataloader_clipreid.py#L170C3-L170C30) to `'dense'` and “batch_size_eval = 30 ” =》 “1”。（I didn't provide "RRS-dense" version directly in the codebase, but it can be implemented by combining the logic of "RRS-test" and "dense".）
This allows you to evaluate all frames, but it will slow down the speed. Ultimately, RRS sampling is a strategy designed to strike a balance between speed and performance.

**Comment by HJoonKwon** [comment]
Thank you for answering the questions!
